---
title: H2AX/Edu Anticolocalization Screen - siKinome - cellHTS2 Data prep
author: "Robin Sebastian/Gianluca Pegoraro"
date: "March 29th 2021"
output: "html_document"
---

### README

This script reads metadata associated with library reformatting, and the data generated by Columbus. Data and metadata are reformatted and joined. The data is then formatted for [CellHTS2](http://www.bioconductor.org/packages/release/bioc/html/cellHTS2.html) analysis.

### Analysis Setup

Load packages.

```{r load-packages}
library(tidyverse)
library(fs)
library(readxl)
library(cellHTS2)
```

### User variables input and settings specification

Set the environment variables

```{r set-variables}
METADATA_DIR<- "hts2_metadata"

IN_DIR <- "hts2_input"

OUT_DIR <- "hts2_output"

WELL_DIR <- "columbus_input"

REFORMAT_DIR <- "reformat_metadata"

GLOB_PAT <- "*result.1.txt"
```

Generate the `r IN_DIR` directory for the well level data to be used in the`cellHTS2` analysis. In addition, create an empty template for the `cellHTS2` description file in the `r METADATA_DIR` directory. **This file contains information about the experiment, and it must be filled in in a text editor before continuing with the analysis. Without this file cellHTS2 won't run**.

```{r make-matadata-dir}
if(!dir_exists(IN_DIR)){
dir_create(IN_DIR)
}

if(!file_exists(file.path(METADATA_DIR,"Description.txt"))) { 
    templateDescriptionFile(filename = "Description.txt", path = METADATA_DIR)
}
```

Set the `knitr` options and the `ggplot2` theme.

```{r set-options}
knitr::opts_chunk$set(
    cache = FALSE,
    fig.path = file.path("knitr_output/"),
    fig.height = 9,
    fig.width = 12,
    message = FALSE,
    warning = FALSE
    )

theme_set(theme_bw())
```

### READ AND PROCESS METADATA

Read the Ambion siRNA library layout information provided by ThermoFisher, rename some variables, eliminate empty rows, and select only relevant columns.

Function to read and wrangle Ambion metadata.

```{r read-ambion-files}
read_ambion <- function(ambion_path){
  dir_ls(path = ambion_path,
         glob = "*.xls") %>%
  map_df(read_excel,
          .id = "File_name") %>%
  select(ambion_barcode = `Plate ID`,
         ambion_well = `Location (Row-Col)`,
         ambion_sample = `Sample ID`,
         ambion_row = Row,
         ambion_col = Col,
         refseq_id = `RefSeq Accession Number`,
         gene_id = `Gene ID`,
         gene_symbol = `Gene Symbol`,
         sirna_id = `siRNA ID`,
         exons = `Exon(s) Targeted`,
         sense = `Sense siRNA Sequence`,
         antisense = `Antisense siRNA Sequence`) %>%
  filter(!is.na(gene_id)) 
}
```

Read the Kinome library Ambion metadata.

```{r read-epi-ambion}
ambion_dir <- file.path(REFORMAT_DIR, "kinome", "ambion_files")

gene_tbl <- read_ambion(ambion_dir)

glimpse(gene_tbl)
```

Sanity check to make sure that the list contains all the Ambion barcodes. 3 Ambion plates only contain 5 siRNA oligos each. The remaining 24 contain 88 genes each(Columns 1 - 11).

```{r}
gene_tbl %>%
  count(ambion_barcode)
```

Read and wrangle the Kinome resuspension metadata output by the PE Janus logs. Resuspend 0.25 nmoles of siRNA oligos per well in 50 ul of water in 96 well plates delivered by Ambion/ThermoFisher. The final siRNA concentration is 5 uM.

```{r read-epi-resuspension}
resuspension_dir <- file.path(REFORMAT_DIR, 
                              "kinome",
                              "reformatting_logs",
                              "resuspension")

resusp_tbl <- dir_ls(resuspension_dir,
                     glob = "*.xlsx") %>%
  map_dfr(read_excel, .id = "file_name") %>%
  select(reservoir_barcode = SrcRack,
         reservoir_well = SrcWell,
         ambion_barcode = Rack,
         ambion_well = Well) %>%
  mutate(ambion_barcode = ambion_barcode %>% str_remove("Q1_001")) %>%
  filter((reservoir_barcode != ambion_barcode) & !(reservoir_barcode %in% c("N/A", "B01")))

glimpse(resusp_tbl)
```

Sanity check to make sure that we still have the proper number of Ambion barcodes. All wells in all Ambion plates are present in this table.

```{r}
resusp_tbl %>%
  count(ambion_barcode)
```

Read and wrangle the Kinome compression logs: transfer 45 ul from a 96-well plate to a Greiner 384 784201 storage plate (Mother plate).

```{r read-epi-compr}
compr_dir <- file.path(REFORMAT_DIR,
                       "kinome",
                       "reformatting_logs",
                       "compression")

compr_tbl <- dir_ls(compr_dir,
                    glob = "*.xlsx") %>%
  map_dfr(read_excel, .id = "file_name") %>%
  select(ambion_barcode = SrcRack,
         ambion_well = SrcWell,
         mother_barcode = Rack,
         mother_well = Well) %>%
  filter((ambion_barcode != mother_barcode) & ambion_barcode != "N/A") %>%
  mutate(ambion_barcode = ambion_barcode %>% str_remove("Q1_001"))

glimpse(compr_tbl)
```

Sanity checks for the proper count of Ambion bracodes and mother plates barcodes.

```{r}
compr_tbl %>%
  count(ambion_barcode)

compr_tbl %>%
  count(mother_barcode)
```

Read and wrangle the Kinome transfer from 384-well Greiner storage plates to 384-well Labcyte LDV plates (Daughter plates) metadata output by the PE Janus logs. This step replicates the library 3X.

```{r read-epi-rep}
rep_dir <- file.path(REFORMAT_DIR,
                     "kinome",
                     "reformatting_logs",
                     "replicates")

rep_tbl <- dir_ls(rep_dir,
                  glob = "*.xlsx") %>%
  map_dfr(read_excel, .id = "file_name") %>%
  select(mother_barcode = SrcRack,
         mother_well = SrcWell,
         rep_barcode = Rack,
         rep_well = Well) %>%
    filter((mother_barcode != rep_barcode) & mother_barcode != "N/A")

glimpse(rep_tbl)
```

Sanity checks for the proper count of Ambion barcodes and mother plates barcodes.

```{r}
rep_tbl %>%
  count(mother_barcode)

rep_tbl %>%
  count(rep_barcode)
```

Read and wrangle the Epigenetics spotting from 384-well LDV Labcyte storage plates to 384-well PE CellCarrier Ultra imaging plates metadata output by the Echo525.

```{r read-epi-imaging}
imaging_dir <- file.path(REFORMAT_DIR,
                           "kinome",
                           "reformatting_logs",
                           "imaging")

imaging_glob <- "*_Transfer_*.csv"


imaging_tbl <- dir_ls(imaging_dir,
                      glob = imaging_glob) %>%
  keep(~ str_detect(.x, "Exceptions",
                    negate = TRUE)) %>%
  read_csv(skip = 8) %>%
  select(rep_barcode = `Source Plate Barcode`,
         rep_well = `Source Well`,
         imaging_barcode = `Destination Plate Barcode`,
         imaging_well = `Destination Well`,
         transfer_vol = `Transfer Volume`,
         actual_vol = `Actual Volume`,
         current_vol = `Current Fluid Volume`
         ) %>%
  # Eliminate transfer exceptions and negative/positive controls in col 23
  filter(!is.na(rep_barcode), 
         str_detect(rep_barcode, "HT"))
```

Sanity checks for the proper count of Ambion barcodes and mother plates barcodes.

```{r}
imaging_tbl %>%
  count(rep_barcode)

imaging_tbl %>%
  count(imaging_barcode)
```

Generate the final annotation table and save is to disk. Select only plates 1 - 6 of the libraries, since plate 7 only has 15 samples and cannot be processed in downstream steps: the B-score cannot be calculated using only 15 samples in one single column. 

```{r}
run_tbl <- tibble(mother_barcode = paste0('HT000', 46:51),
                  plate_number = 1:6)

annotation_tbl <- imaging_tbl %>% 
  left_join(rep_tbl, by = c("rep_barcode", "rep_well")) %>%
  arrange(mother_barcode, mother_well) %>%
  left_join(compr_tbl, by = c("mother_barcode", "mother_well")) %>%
  left_join(gene_tbl, by = c("ambion_barcode", "ambion_well")) %>%
  left_join(run_tbl, by = "mother_barcode") %>%
  mutate(mother_row = str_extract(mother_well, "[A-Z]"),
         mother_row_num = match(mother_row, LETTERS),
         mother_col = as.numeric(str_extract(mother_well, "[0-9]+")),
         rep_row = str_extract(rep_well, "[A-Z]"),
         rep_row_num = match(rep_row, LETTERS),
         rep_col = as.numeric(str_extract(rep_well, "[0-9]+")),
         imaging_row = str_extract(imaging_well, "[A-Z]"),
         imaging_row_num = match(imaging_row, LETTERS),
         imaging_col = as.numeric(str_extract(imaging_well, "[0-9]+"))) %>%
  select(imaging_barcode, imaging_well, imaging_row,
         imaging_row_num, imaging_col, transfer_vol,
         actual_vol, plate_number, rep_barcode,
         rep_well, rep_row, rep_row_num,
         rep_col, mother_barcode,
         mother_well, mother_row, mother_row_num,
         mother_col, ambion_barcode:sense) %>%
  filter(!is.na(plate_number))


write_csv(annotation_tbl, 
          file.path("knitr_output", 
                    "kinome_annotation_tbl.csv"))

glimpse(annotation_tbl)
```

Check number of wells for each mother plate.

```{r}
annotation_tbl %>% 
  count(mother_barcode)
```

Plot the siRNA layout for the mother plates.

```{r mother-layout, echo = FALSE}
annotation_tbl %>%
  ggplot(aes(x = mother_col,
             y = mother_row_num, 
             label = sirna_id)) +
  geom_text(size = 0.8) +
  scale_y_reverse(breaks = 1:16, labels = LETTERS[1:16]) +
  scale_x_continuous(breaks = 1:24) +
  facet_wrap(vars(mother_barcode)) +
  xlab("Column") +
  ylab("Row") +
  ggtitle("Mother Plates Layout")
```

Plot the siRNA layout for the imaging plates.

```{r fig.imaging-layout}
annotation_tbl %>%
  ggplot(aes(x = imaging_col, 
             y = imaging_row_num, 
             label = sirna_id)) +
  geom_text(size = 0.8) +
  scale_y_reverse(breaks = 1:16, labels = LETTERS[1:16]) +
  scale_x_continuous(breaks = 1:24) +
  facet_wrap(vars(imaging_barcode)) +
  xlab("Column") +
  ylab("Row") +
  ggtitle("Imaging Plates Layout")
```

### Read Data Generated by Columbus

Read the well-level data generated by Columbus. The well level data must be in the `WELL_DIR` and the file names *must match* the `r GLOB_PAT` `glob` pattern.

```{r read-columbus-data}
well_tbl <- dir_ls(path = WELL_DIR,
                    recurse = T,
                    glob = GLOB_PAT) %>%
    read_tsv(id = "file_name",
             col_types = list(`Spots Edu Pos - center_border_ratio - Median per Well` = "n")) %>%
    select(plate_index = PlateName,
          well_index = WellName,
          row = Row,
          column = Column,
          nuclei_n = `Nuclei Final - Number of Objects`,
          nuclei_area = `Nuclei Final - Nucleus Area [µm²] - Mean per Well`,
          nuclei_round = `Nuclei Final - Nucleus Roundness - Mean per Well`,
          nuclei_mean_edu = `Nuclei Final - Intensity Nucleus BP676/29 Mean - Mean per Well`,
          nuclei_H2AX_spots_per_nucleus = `Nuclei Final - Number of Spots - Mean per Well`,
          nuclei_mean_edu_pos = `Nuclei Final - Nuclei Final Edu Pos - Mean per Well`,
          spots_edu_pos_n = `Spots Edu Pos - Number of Objects`,
          spots_edu_pos_median_ratio = `Spots Edu Pos - center_border_ratio - Median per Well`)

write_csv(well_tbl, file = file.path("knitr_output", 
                                     "edu_kinome_well_results.csv"))

glimpse(well_tbl)
```

## `cellHTS2` data wrangling

Create input and output folder hierarchies in the `r IN_DIR` and `r OUT_DIR` directories, respectively.

```{r create-hts2-directories}
well_tbl %>%
  select(nuclei_n:spots_edu_pos_median_ratio) %>%
  names() %>% walk(~ dir_create(file.path(IN_DIR, .x)))

well_tbl %>%
  select(nuclei_n:spots_edu_pos_median_ratio) %>%
  names() %>% walk(~ dir_create(file.path(OUT_DIR, .x)))
```

### Annotation

Make the annotation file for `cellHTS2`, name it `Annotation.txt` and save it to the in the `r IN_DIR` directory.

```{r make-annotation-table}
annotation_hts2_tbl <- annotation_tbl %>%
  select(
    Plate = plate_number,
    Well = imaging_well,
    GeneID = gene_id,
    GeneSymbol = gene_symbol,
    siRNAID = sirna_id
  ) %>%
  distinct() %>%
  mutate(Well = sprintf("%s%02d",
                        str_extract(Well, "\\w"),
                        as.numeric(str_extract(Well, "\\d+")))) %>%
  arrange(Plate, Well)

glimpse(annotation_hts2_tbl)

write_tsv(annotation_hts2_tbl,
          file = file.path(METADATA_DIR, "Annotation.txt"))
```

### Plate Configuration

Generate the configuration file `Plateconf.txt` for `cellHTS2` according to the documentation specifications: \> The software expects this to be a rectangular table in a tabulator delimited text file, with mandatory columns Plate, Well, Content, plus two additional header lines that give the total number of wells and plates (see Table ?? for an example). The content of this file (except the two header lines) are stored in slot plateConf of x. As the name suggests, the Content column provides the content of each well in the plate (here referred to as the well annotation). Mainly, this annotation falls into four categories: empty wells, wells targeting genes of interest, control wells, and wells containing other things that do not fit in the previous categories. The first two types of wells should be indicated in the Content column of the plate configuration file by empty and sample, respectively, while the last type of wells should be indicated by other. The designation for the control wells in the Content column is more flexible. By default, the software expects them to be indicated by pos (for positive controls), or neg (for negative controls). However, other names are allowed, given that they are specified by the user whenever necessary (for example, when calling the writeReport function). This versatility for the control wells' annotation is justified by the fact that, sometimes, multiple positive and/or negative controls can be employed in a given screen, making it useful to give different names to the distinct controls in the Content column. More- over, this versatility might be required in multi-channel screens for which we frequently have reporter-specific controls. The Well column contains the name of each well of the plate in alphanu- meric format (in this case, A01 to P24), while column Plate gives the plate number (1, 2, ...). These two columns are also allowed to contain regular expressions. In the plate configuration file, each well and plate should be covered by a rule, and in case of multiple definitions only the last one is considered. For example, in the file shown in Table ??, the rule specified by the first line after the column header indicates that all of the wells in each of the 57 assay plate contain "sample". However, a following rule indicate that the content of wells A01, A02 and B01 and B02 differ from "sample", containing other material (in this case, "other" and controls). Note that the well annotations mentioned above are used by the software in the normalization, quality control, and gene selection calculations. Data from wells that are annotated as empty are ignored, i. e. they are set to NA.

The configuration file tells `cellHTS2` where the samples, controls and empty wells are on each plate. The first 6 lines of the file (Headers and controls positions) are hardcoded for now. The samples layout are is appended from the processed information obtained from the Janus logs.

```{r make-config-table}
line1 <- "Wells: 384"
line2 <- "Plates: 6"
line3 <- "Plate\tWell\tContent"
line4 <- "*\t*\tempty"

odd_rows <- paste(LETTERS[seq(1, 16, by = 2)], collapse = ",")
even_rows <- paste(LETTERS[seq(2, 16, by = 2)], collapse = ",")

line5 <-
  paste0("*\t[", odd_rows, "]23\tneg") # siScramble in Column 23 odd rows (A, C, ..., O)
line6 <-
  paste0("*\t[", odd_rows, "]24\tpos") # siRIF1 in Column 24 odd rows (A, C, ..., O)
line7 <-
  paste0("*\t[", even_rows, "]23\tpos2") # siPLK1 in Column 23 even rows (B, D, ..., P)

header <-  c(line1, line2, line3, line4, line5, line6, line7)

header

# Write header to file
conf <- file(file.path(METADATA_DIR, "Plateconf.txt"), "w")
writeLines(header, conf)
close(conf)
```

The samples layout is appended from the processed information obtained from the Janus and Echo 525 logs and is written out to the the `r METADATA_DIR` directory.

```{r append-config-table}
config_tbl <- annotation_hts2_tbl %>%
  filter(!is.na(GeneID)) %>%
  select(Plate, Well) %>%
  mutate(Content = "sample")

glimpse(config_tbl)

write_tsv(config_tbl,
          file = file.path(METADATA_DIR, "Plateconf.txt"),
          append = T) # append 
```

### Measurements

Generate a `Master_Barcode_List.txt` file and save it in the working directory (i.e. the same directory where the .Rmd file is). According to the `cellHTS2` specifications, the file should have these columns: `PlateName` (The plate barcode, as determined in the Janus log files, a string), `Plate` (The plate number in the library, a number), `Replicate` (Self explanatory, a number) and `Batch` (if the experiment or replicate was run in different batches, not necessary). This file contains the information on which plates have been analyzed, and on what the internal organization of the run is. The `Master_Barcode_List.txt` file is used to generate the measurement files to be read by `cellHTS2`.

```{r read-barcode-list}
barcodes_tbl <- read_tsv(file.path(METADATA_DIR, "Master_Barcode_List.txt"))

glimpse(barcodes_tbl)
```

### Plate List

Generate and write the cellHTS2 `Platelist.txt` file in the `r IN_DIR` directory.

```{r makePlatelist}
well_tbl %>%
  left_join(barcodes_tbl, by = c("plate_index" = "PlateName")) %>%
  pivot_longer(cols = nuclei_n:spots_edu_pos_median_ratio,
               names_to = "measurement",
               values_to = "value") %>%
  group_by(Plate, Replicate, Batch, measurement, plate_index) %>%
  group_keys() %>%
  mutate(Filename = paste0(plate_index, "_cellHTS2.txt")) %>%
  group_by(measurement) %>%
  group_walk(~ write_tsv(
    select(.x, Filename, Plate, Replicate, Batch),
    file.path(IN_DIR,
              .y$measurement,
              "Platelist.txt")
  ))

dir_info(path = IN_DIR,
         glob = "*.txt",
         recurse = T) %>% select(1:2)
```

Reformat the well names according to cellHTS2 specifications and write out one cellHTS2 measurement files on a per measurement, per plate basis in the `r IN_DIR` directory.

```{r formatWellsHTS2Format}
meas_tbl <- left_join(well_tbl, barcodes_tbl, by = c("plate_index" = "PlateName")) %>%
  mutate(well = sprintf("%s%02d",
                        str_extract(well_index, "\\w"),
                        as.numeric(str_extract(well_index, "\\d+"))))

meas_tbl %>% pivot_longer(cols = nuclei_n:spots_edu_pos_median_ratio,
                          values_to =  "Value",
                          names_to = "measurement") %>%
             group_by(measurement, plate_index) %>%
             group_walk( ~ write_tsv(
                select(.x, Plate, WellName = well_index, Value),
                file.path(IN_DIR,
                          .y$measurement,
                          paste0(.y$plate_index,
                                 "_cellHTS2.txt")),
                col_names = F,
                na = "NaN"
              ))

dir_info(path = IN_DIR,
         glob = "*_cellHTS2.txt",
         recurse = T) %>% select(1:2)
```

Document the information about the analysis session

```{r sessionInfo, include=TRUE, echo=TRUE, results='markup'}
sessionInfo()
```
